plugins/gs2.c

```diff
@@ -1073,6 +1096,13 @@ gs2_verify_initial_message(context_t *text,
         if (ret != SASL_OK)
             return ret;
 
+        if (sparams->cbinding == NULL ||
+            strcmp(sparams->cbinding->name, text->cbindingname) != 0) {
+            /* Try to get desired channel binding data */
+            gs2_server_get_channel_binding(sparams,
+                                           text->plug.server->mech_name,
+                                           text->cbindingname);
```

Comment:
Overall the patch looks sensible, except for one small thing: you don't test the result of this call and you should fail authentication if it doesn't return SASL_OK.

Reply:
Thanks for the review. Good point! Indeed the error handling is not smooth here and I cannot remember my initial intention.
Obviously I wanted to proceed with the default error handling of the challenge/response protocol when the client uses a wrong or unexpected channel binding type.
Currently I'm not sure if it is a good idea to abort the challenge/response protocol when the server receives an unexpected text->cbindingname. This is the current behavior of previous SASL implementations. I will think about a better solution when I add a client/server example.
What's your preference?
1. Continue with default error handling when client uses an unexpected channel binding type. If yes, I change the function declaration in static ***void*** gs2_server_get_channel_binding(...) and just ignore the callback result. You will find similar coding in lib/server.c (e.g. getopt).
2. Abort the challenge/response protocol when client uses an unexpected channel binding type or callback returns an error.

Reply:
gs2_server_get_channel_binding now ignores the callback result. Error handling is done by protocol if callback cannot set the correct channel binding data.